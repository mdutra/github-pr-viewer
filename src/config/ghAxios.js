import axios from 'axios';

// Axios intance to access GitHub API v3
const ghAxios = axios.create({
  baseURL: 'https://api.github.com/',
  headers: {
    // Set version
    Accept: 'application/vnd.github.v3+json',
  },
});

export default ghAxios;
