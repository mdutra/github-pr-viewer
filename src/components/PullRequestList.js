import React, { Component } from 'react';
import PropTypes from 'prop-types';

import ghAxios from '../config/ghAxios';
import CustomTable from './CustomTable';

const PER_PAGE = 10;

class PullRequestList extends Component {
  constructor(props) {
    super(props);

    this.state = {
      data: [],
      page: 1,
      isLoading: true,
      error: undefined,
    };

    this.handleOnChangePage = this.handleOnChangePage.bind(this);
  }

  fetchPullRequests(page, per_page) {
    const { user, repo } = this.props.match.params;

    return ghAxios.get(`repos/${user}/${repo}/pulls`, {
      params: {
        state: 'open',
        page,
        per_page,
      },
    });
  }

  async componentDidMount() {
    try {
      const { data } = await this.fetchPullRequests(
        this.state.page,
        // Fetch 2 pages + 1 row to make pagination smoother
        PER_PAGE * 2 + 1,
      );

      this.setState({
        data: data.map(({ title }) => ({ title })),
        isLoading: false,
      });
    } catch (e) {
      this.setState({
        isLoading: false,
        error: e.message,
      });
    }
  }

  async handleOnChangePage(page) {
    const nextPage = page + 1;
    const currentPage = this.state.page;

    // If going a page forward
    if (nextPage > currentPage) {
      const { data } = await this.fetchPullRequests(nextPage, PER_PAGE);

      this.setState({
        data: [
          ...this.state.data,
          ...data.map(({ title, body }) => ({ title, body })),
        ],
        page: nextPage,
      });
    } else {
      this.setState({ page: nextPage });
    }
  }

  render() {
    const { user, repo } = this.props.match.params;

    return (
      <CustomTable
        repo={user + '/' + repo}
        data={this.state.data}
        isLoading={this.state.isLoading}
        pageSize={PER_PAGE}
        onChangePage={this.handleOnChangePage}
        emptyDataSourceMessage={this.state.error}
      />
    );
  }
}

PullRequestList.propTypes = {
  match: PropTypes.shape({
    params: PropTypes.shape({
      user: PropTypes.string.isRequired,
      repo: PropTypes.string.isRequired,
    }).isRequired,
  }).isRequired,
};

export default PullRequestList;
