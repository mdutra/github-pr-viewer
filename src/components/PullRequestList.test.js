import React from 'react';
import { render, cleanup, wait } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import ShallowRenderer from 'react-test-renderer/shallow';

import PullRequestList from './PullRequestList';
import ghAxios from '../config/ghAxios';

jest.mock('../config/ghAxios');

afterEach(cleanup);

it("displays message when there's no PR to show", async () => {
  const user = 'test';
  const repo = 'test';

  ghAxios.get.mockResolvedValueOnce({
    data: [],
  });

  const { container } = render(
    <PullRequestList match={{ params: { user, repo } }} />,
  );

  await wait(expect(container).toBeInTheDocument());

  expect(container).toHaveTextContent('No records to display');
});

it('matches snapshot', async () => {
  const renderer = new ShallowRenderer();
  const user = 'test';
  const repo = 'test';

  const result = renderer.render(
    <PullRequestList match={{ params: { user, repo } }} />,
  );

  expect(result).toMatchSnapshot();
});
