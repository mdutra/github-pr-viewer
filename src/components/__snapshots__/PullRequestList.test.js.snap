// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<CustomTable
  data={Array []}
  emptyDataSourceMessage="No records to display"
  isLoading={true}
  onChangePage={[Function]}
  pageSize={10}
  repo="test/test"
/>
`;
